### PR TITLE
Add GitHub Actions workflow to create a PR to promote nightly builds

### PR DESCRIPTION
**Pull request by danielrozenberg**

### CODE REVIEW

File: workflows/promote-job.ts
Comment (estherkim) [inline workflows/promote-job.ts:7]: Can these be combined into 1 line? having BaseOctokit and Octokit and octokit is confusing to read

Comment (estherkim) [reply workflows/promote-job.ts:7]: Actually I dislike having another npm package when you can easily create a PR with octokit/rest https://octokit.github.io/rest.js/v18#pulls-create

File: workflows/promote-job.ts
Comment (estherkim) [inline workflows/promote-job.ts:64]: update name

File: workflows/promote-job.ts
Comment (estherkim) [inline workflows/promote-job.ts:86]: how is this a valid head value?

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts:7]: The plugin lets me create a PR with file changes. The method you linked only creates a PR, but that would mean I would have to modify the file inline, create a commit, push a branch, and create a PR manually - the plugin consolidates all of these into a single command

fwiw this plugin was created by [gr2m](https://github.com/gr2m) who is the maintainer of the Octokit JS repo, so if we're  using Octokit it's probably fine to use this plugin.

wdyt?


As for consolidating the two lines together, this is what it might end up looking like. Not sure it's any cleaner...
```suggestion
import {Octokit} from '@octokit/rest';

const octokit = new (Octokit.plugin(createPullRequest))({
  auth: process.env.GITHUB_TOKEN,
});
```

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts:86]: This is the new branch name to create the PR with: https://www.npmjs.com/package/octokit-plugin-create-pull-request#:~:text=head%3A%20%22pull%2Drequest%2Dbranch%2Dname%22%2C

File: .github/workflows/promote-reusable-workflow.yml
Comment (danielrozenberg) [reply .github/workflows/promote-reusable-workflow.yml]: Done d2f4731

File: .github/workflows/promote-reusable-workflow.yml
Comment (danielrozenberg) [reply .github/workflows/promote-reusable-workflow.yml]: Done fb3f591

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts]: Done fb3f591

File: .github/workflows/promote-reusable-workflow.yml
Comment (danielrozenberg) [reply .github/workflows/promote-reusable-workflow.yml]: I blindly copied this from workflows in amphtml, so I don't know if it is
Removed it in 2be44ae, let's see if anything breaks

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts]: We don't! Good nit :D 5369d03

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts]: Done 8bd7813

File: .github/workflows/promote-nightly.yml
Comment (estherkim) [inline .github/workflows/promote-nightly.yml:6]: Is this the right trigger? It is being run when you push commits to this PR

File: workflows/promote-job.ts
Comment (estherkim) [inline workflows/promote-job.ts:26]: I don't think we need timing when it's available in the GH workflow

File: workflows/promote-job.ts
Comment (estherkim) [reply workflows/promote-job.ts:7]: I think the suggested change looks cleaner! And LGTM about using the plugin.

File: tsconfig.json
Comment (estherkim) [inline tsconfig.json:2]: nit: can it be named `scripts/` instead? open to suggestions

File: workflows/promote-job.ts
Comment (estherkim) [inline workflows/promote-job.ts:2]: Colors don't show up and it would make this file simpler if we took it out of the logging.

File: workflows/promote-nightly.ts
Comment (estherkim) [inline workflows/promote-nightly.ts:16]: Do we need the `04` prefix or can we do a replace on the last 13? It would help in not having to worry about the prefixes for all the channels.

File: workflows/promote-nightly.ts
Comment (estherkim) [inline workflows/promote-nightly.ts:18]: The title and body could probably be parameterized. We already have `channel-name` as an input on the workflow - it could be used in `promote-job.ts`

File: workflows/promote-job.ts
Comment (estherkim) [reply workflows/promote-job.ts:86]: I see, can it be changed to `promote-${CHANNEL_NAME}-${AMP_VERSION}`? Using GITHUB_RUN_ID will work but won't be helpful when we look at branches

File: workflows/promote-nightly.ts
Comment (estherkim) [inline workflows/promote-nightly.ts:10]: TIL the void operator. We don't want to return undefined right? It would be better to return the promise

File: .github/workflows/promote-reusable-workflow.yml
Comment (estherkim) [inline .github/workflows/promote-reusable-workflow.yml:40]: I would add `.ts` to the end of `ts-file` value, this looks funny

File: .github/workflows/promote-reusable-workflow.yml
Comment (estherkim) [inline .github/workflows/promote-reusable-workflow.yml:29]: Does this need to be specified?

File: .github/workflows/promote-reusable-workflow.yml
Comment (estherkim) [inline .github/workflows/promote-reusable-workflow.yml:42]: nit: I think this should be a script arg, not an env var

File: workflows/promote-nightly.ts
Comment (estherkim) [reply workflows/promote-nightly.ts:18]: We could probably get rid of this file alltogether

**Review by estherkim — COMMENTED** [review]
More comments!

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts]: Done db4ed6e

File: tsconfig.json
Comment (danielrozenberg) [reply tsconfig.json]: Done 06f86bc - don't have strong feelings about this directory name

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts]: Done b1efb9d

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts]: Done fbf646c

File: workflows/promote-nightly.ts
Comment (danielrozenberg) [reply workflows/promote-nightly.ts]: We have the JSON verifier step in the CI, and these numbers aren't going to change so it's an easy explicit solution. There _could_ be a situation where we disable nightly temporarily (manually setting it to `null`) in which case a S&R won't work

File: workflows/promote-nightly.ts
Comment (danielrozenberg) [reply workflows/promote-nightly.ts]: Not really, different promote workflows will have different `versionMutator` functions (e.g., the opt-in job on Tuesdays will have to take the existing AMP version from Nightly and set it to the opt-in channels). This is the most generic function I could think of atm

File: workflows/promote-job.ts
Comment (danielrozenberg) [reply workflows/promote-job.ts]: Done 9b3336d

Had to add another field to the `versionsMutator` return value to pass the branch name

File: .github/workflows/promote-nightly.yml
Comment (danielrozenberg) [reply .github/workflows/promote-nightly.yml:6]: `repository_dispatch` allows us to trigger this workflow using the GitHub API:
* https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#repository_dispatch
* https://pakstech.com/blog/github-actions-repository-dispatch/

This did make me realize that I also need to add the input params to the `workflow_dispatch`. Done 2be9e77

File: workflows/promote-nightly.ts
Comment (danielrozenberg) [reply workflows/promote-nightly.ts]: Oh yeah I should've mentioned that. Without the `void` operator I get an eslint error:
```
Promises must be awaited, end with a call to .catch, end with a call to .then
with a rejection handler or be explicitly marked as ignored with the `void`
operator. eslint@typescript-eslint/no-floating-promises
```

We can't await on the top level without jumping through a lot of hoops, but this is okay because Node will wait until all Promises are resolved/rejected before exiting, so there's no need to await this explicitly (though I prefer if we could do that...)

File: .github/workflows/promote-reusable-workflow.yml
Comment (danielrozenberg) [reply .github/workflows/promote-reusable-workflow.yml]: Done 8213fab

File: .github/workflows/promote-reusable-workflow.yml
Comment (danielrozenberg) [reply .github/workflows/promote-reusable-workflow.yml:29]: Eh, not necessarily but it would make things faster once this repo has thousands of commits (e.g., if we commit at least once every weekday, within 2 years there'll be ~500 commit in the repo)

File: .github/workflows/promote-reusable-workflow.yml
Comment (danielrozenberg) [reply .github/workflows/promote-reusable-workflow.yml]: I made this change in 7a27fef - promote-reusable-workflow.yml now has an `extra-args` input field instead of `amp-version`, wdyt?

File: .github/workflows/promote-nightly.yml
Comment (estherkim) [reply .github/workflows/promote-nightly.yml:6]: You also need to prevent these runs from triggered every time you push a commit to this PR - https://github.com/ampproject/cdn-configuration/actions/workflows/promote-nightly.yml

Comment (estherkim) [reply .github/workflows/promote-nightly.yml:6]: (that was my initial point)

File: .github/workflows/promote-reusable-workflow.yml
Comment (estherkim) [reply .github/workflows/promote-reusable-workflow.yml:30]: This action already fetches a single commit by default.

File: .github/workflows/promote-nightly.yml
Comment (danielrozenberg) [reply .github/workflows/promote-nightly.yml:6]: Oh, looks like I just had a syntax error there. It doesn't actually run this workflow on the PR, it just marks this as if it did because it's a PR check. Confusing UI is confusing.

Fixed in 077ab48

File: .github/workflows/promote-reusable-workflow.yml
Comment (danielrozenberg) [reply .github/workflows/promote-reusable-workflow.yml]: Oh I didn't realize that. Done 9330fe8

File: workflows/promote-nightly.ts
Comment (estherkim) [reply workflows/promote-nightly.ts]: I was picturing the opt-in workflow to have an additional step - getting the amp version of the current nightly with an optional override - and then calling the promote script.

That would allow the promote script be the same across workflows and I imagine it would be easier to debug if it was split in different steps.

File: workflows/promote-nightly.ts
Comment (estherkim) [reply workflows/promote-nightly.ts]: Would this work?
```
function runPromoteJob() {

  ...
  
  return createVersionsUpdatePullRequest()
}

runPromoteJob();
```

File: workflows/promote-nightly.ts
Comment (estherkim) [reply workflows/promote-nightly.ts]: It would be simpler if the prefix weren't carried over. The json schema CI check is perfect in catching bad S&Rs.

Do you see disabling nightly as a likely scenario? It would fail the same CI check anyways. And an null check in the code is cleaner to me than having prefixes in places other than `versions.json`

File: workflows/promote-nightly.ts
Comment (danielrozenberg) [reply workflows/promote-nightly.ts]: No, because that creates a function that returns a Promise, and then we just have the same problem :)

Comment (estherkim) [reply workflows/promote-nightly.ts]: Ok then! void it is

File: .github/workflows/promote-nightly.yml
Comment (estherkim) [inline .github/workflows/promote-nightly.yml:22]: Have you tried running this on your own fork? I don't see `jobs.<job_id>.with` as a valid option on https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions

File: .github/workflows/promote-reusable-workflow.yml
Comment (estherkim) [reply .github/workflows/promote-reusable-workflow.yml]: Why extra-args instead? I was picturing something like 
`ts-node script.ts --amp-version=${{ github.event.inputs.amp-version }}`

File: .github/workflows/promote-nightly.yml
Comment (danielrozenberg) [reply .github/workflows/promote-nightly.yml:22]: I haven't tried running it locally, figured we can try it after we merge :P
https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idwith

Comment (estherkim) [reply .github/workflows/promote-nightly.yml:22]: Thanks for the link. Can you test locally? Would help with so much of our back-and-forths today...

File: .github/workflows/promote-reusable-workflow.yml
Comment (danielrozenberg) [reply .github/workflows/promote-reusable-workflow.yml]: Hmm... I guess I didn't think this though. I thought that only the nightly should have one (because the others will default to promoting from one channel to another without explicitly being told which one), but you're right - every promote workflow should have the option to get an --amp_version passed into it.

Done a4bee1f

File: .github/workflows/promote-nightly.yml
Comment (danielrozenberg) [reply .github/workflows/promote-nightly.yml:22]: I can try, see how involved it might be to test it on a separate repo - if it's gonna take >1h I prefer to just merge this PR and fix anything that goes wrong here

Comment (danielrozenberg) [reply .github/workflows/promote-nightly.yml:22]: (will keep you posted)

Comment (estherkim) [reply .github/workflows/promote-nightly.yml:22]: Hopefully not too involved. I do it myself for the workflows I write. Reviewing this PR has taken a lot of my own time today. Thank you!

Comment (danielrozenberg) [reply .github/workflows/promote-nightly.yml:22]: Tested this on my own fork as you suggested and I'm so glad that you did, because it broke in all the places I didn't expect it to break! 5138769 fixes everything that I learned from the fork tests:
* Logs for the fork's successful workflow run: https://github.com/danielrozenberg/cdn-configuration/runs/4732259272?check_suite_focus=true#step:5:1
* PR that was auto generated: https://github.com/danielrozenberg/cdn-configuration/pull/1

File: .github/workflows/promote-nightly.yml
Comment (danielrozenberg) [reply .github/workflows/promote-nightly.yml]: Ha, looks like `workflow_dispatch` is enough! I don't know why `repository_dispatch` exists if you can trigger workflows with Octokit. Removed in eb408cd

File: workflows/promote-nightly.ts
Comment (danielrozenberg) [reply workflows/promote-nightly.ts]: Take a look at #36 (still in draft) and lmk if you still think parameterizing this would make it look better (the answer might be yes! You decide)

**Review by estherkim — APPROVED** [review]
LGTM to unblock

**Pull request merged**
